src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -137,17 +132,33 @@ public ICollection<string> Keys
         {
             get
             {
-                EnsureWritable();
-                return InnerDictionary.Keys;
+                Upgrade();
```

Comment:
Plus same comment in other similar scenarios. (e.g. generic Keys property below)

---
In reply to: [60490919](https://github.com/aspnet/Routing/pull/309#discussion_r60490919) [](ancestors = 60490919)